pyo3-derive-backend/src/frompy.rs

```diff
@@ -0,0 +1,428 @@
+    }
+}
+
+/// Get the extraction function that's called on the input object.
+///
+/// Valid arguments are `get_item`, `get_attr` which are called with the
+/// stringified field identifier or a function call on `PyAny`, e.g. `get_attr("attr")`
+fn ext_fn(attrs: &[Attribute], field_ident: &Ident) -> Result<syn::ExprCall> {
+    let attr = if let Some(attr) = attrs.iter().find(|a| a.path.is_ident("extract")) {
+        attr
+    } else {
+        return Ok(parse_quote!(getattr(stringify!(#field_ident))));
+    };
+    if let Ok(ident) = attr.parse_args::<Ident>() {
+        if ident != "getattr" && ident != "get_item" {
+            Err(syn::Error::new_spanned(
+                ident,
+                "Only get_item and getattr are valid for extraction.",
+            ))
+        } else {
+            let arg = field_ident.to_string();
+            Ok(parse_quote!(#ident(#arg)))
+        }
+    } else if let Ok(call) = attr.parse_args() {
+        Ok(call)
+    } else {
+        Err(syn::Error::new_spanned(
+            attr,
+            "Only get_item and getattr are valid for extraction,\
+            both can be passed with or without an argument, e.g. \
+            #[extract(getattr(\"attr\")] and #[extract(getattr)]",
+        ))
+    }
+}
+
+/// Returns the name of the variant for the error message if no variants match.
+fn maybe_renamed_err(attrs: &[syn::Attribute]) -> Result<Option<syn::LitStr>> {
+    for attr in attrs {
+        if !attr.path.is_ident("rename_err") {
+            continue;
+        }
+        let attr = attr.parse_meta()?;
+        if let syn::Meta::NameValue(nv) = &attr {
+            match &nv.lit {
+                syn::Lit::Str(s) => {
+                    return Ok(Some(s.clone()));
+                }
+                _ => {
+                    return Err(syn::Error::new_spanned(
+                        attr,
+                        "rename_err attribute must be string literal: #[rename_err=\"Name\"]",
+                    ))
+                }
+            }
+        }
+    }
+    Ok(None)
+}
+
+fn verify_and_get_lifetime(generics: &syn::Generics) -> Result<Option<&syn::LifetimeDef>> {
+    let lifetimes = generics.lifetimes().collect::<Vec<_>>();
+    if lifetimes.len() > 1 {
+        return Err(syn::Error::new_spanned(
+            &generics,
+            "Only a single lifetime parameter can be specified.",
+        ));
+    }
+    Ok(lifetimes.into_iter().next())
+}
+
+/// Derive FromPyObject for enums and structs.
+///
+///   * Max 1 lifetime specifier, will be tied to `FromPyObject`'s specifier
+///   * At least one field, in case of `#[transparent]`, exactly one field
+///   * At least one variant for enums.
+///   * Fields of input structs and enums must implement `FromPyObject`
+///   * Derivation for structs with generic fields like `struct<T> Foo(T)`
+///     adds `T: FromPyObject` on the derived implementation.
+pub fn build_derive_from_pyobject(tokens: &mut DeriveInput) -> Result<TokenStream> {
```

Comment:
I'm not sure I've spotted a case where tokens is modified.

```suggestion
pub fn build_derive_from_pyobject(tokens: &DeriveInput) -> Result<TokenStream> {
```